Tree rewiring

svouse opened the pull request:
More project-focused implementation, mainly rewiring the tree.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/101)
<!-- Reviewable:end -->

svouse committed:
added getTreeJson methods

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into model_implementation

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into model_implementation

svouse committed:
filling in project removal methods

svouse committed:
getTreeJson added to resource_set.js, fixed comments

svouse committed:
changed removal to type, fixed typo

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into model_implementation

svouse committed:
responded to comments

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into model_implementation

svouse committed:
more comment response

svouse committed:
beginning view method updates

svouse committed:
continued adding methods

svouse committed:
Merge branch 'develop' of https://github.com/google/blockly-devtools into model_implementation

svouse committed:
fixed merge conflicts

svouse committed:
fixes for PR

svouse committed:
fixes for PR

svouse committed:
comment responses

svouse committed:
comment responses

svouse committed:
comment responses

svouse committed:
fixed addBlockDefinition

svouse committed:
more work

svouse committed:
removed duplicate method, extra space

svouse committed:
comment responses

svouse committed:
Merge branch 'model_implementation' into more_model_implementation

svouse committed:
pr fixes

svouse committed:
fixed conflicts

svouse committed:
added JSON from array method to resource for ease of use

svouse committed:
merge conflict fixes

svouse committed:
work on 7/10/17 TODOs

svouse committed:
work on 7/10/17 TODOs

svouse committed:
fixed merge conflicts with #97

svouse committed:
fixed accidental removal of menu formatting

svouse committed:
continued fixes

svouse committed:
continued rewiring work

svouse committed:
preparing for pr

svouse committed:
reshuffling, changing of passed variable types

svouse committed:
merge conflict fixes

svouse committed:
mistake fixes

svouse committed:
removed note; we have 99 issues and tab switching is one

svouse committed:
comment fixes

svouse committed:
added TODOs

svouse committed:
comment update

svouse committed:
small update to JSDoc

chynu commented:
Review status: 0 of 11 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/project_controller.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyL62c7s0-fVQL1zi1:-KoyL62c7s0-fVQL1zi2:b-58a7la) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L91)):*
> ```JavaScript
>    * @param {string} toolboxName Name of the toolbox to add to the project.
>    */
>   //TODO #100: have component addition return a boolean and have tree addition depend on it
> ```

Good idea - another suggestion is possibly returning the object you just added (and if it wasn't successfully added, return `null`)? It may be more useful to have the object as well as an indication of whether an object was successfully added (in case there is another operation we would like to do right after adding). It has the advantages of returning a boolean in addition to more information.

---

*[src/controller/project_controller.js, line 92 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyKi99R0ZpWVthYuzo:-KoyKi99R0ZpWVthYuzp:b-lrw87g) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L92)):*
> ```JavaScript
>    */
>   //TODO #100: have component addition return a boolean and have tree addition depend on it
>   addToolbox(toolboxName) {
> ```

Would it be a better idea to pass in the toolbox itself to add when adding? (As well as in the other add functions below.)

Update: Either works :D

---

*[src/model/project.js, line 330 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyMPbX0Fl6_X2-Zcrm:-KoyMPbX0Fl6_X2-Zcrn:b-kd2sdj) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/project.js#L330)):*
> ```JavaScript
>    * @return {!Object} The JSON for the tree's library section.
>    */
>   librarySetJson() {
> ```

`getLibrarySetJson()`? (here and below)

(for naming consistency, i suppose)

Also, perhaps it would be better design to put these json generating functions in their corresponding classes (i.e. instead of here, there might be a `getJson()` in every set class?).

---

*[src/model/resource.js, line 90 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyPBfMuQSlRScraWCH:-KoyPBfMuQSlRScraWCI:b-5lnqnc) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/resource.js#L90)):*
> ```JavaScript
>    */
>   makeTreeJsonFromArray(nameArray) {
>     const treeJson;
> ```

I believe JavaScript doesn't let you initialize variables as `const` if you don't give it a value upon naming it. In cases like this I'm pretty sure we have to use `let` or `var`. (Or at least my linter complains about it)

---

*[src/model/resource_set.js, line 57 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyPXBlOe4X1T3aQEKI:-KoyPXBlOe4X1T3aQEKJ:bf9sg6) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/resource_set.js#L57)):*
> ```JavaScript
>   /**
>    * Adds a resource to the set.
>    * @param {!Object} resource The resource to be added.
> ```

Nit - `{!Resource}`? Not sure how JSdoc works but they're all extensions of it, so maybe that works?

---

*[src/model/toolbox_set.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyTttre8roj1uxTYeW:-KoyTttre8roj1uxTYeX:bhb4qdb) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/toolbox_set.js#L91)):*
> ```JavaScript
>    * @param {string} toolboxName The name of the toolbox to be cleared.
>    */
>   clearToolbox(toolboxName) {
> ```

Would it be better to have this method in the `Toolbox` class itself? 0:

---

*[src/model/workspace_configuration_set.js, line 78 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQAjmCeXue_RPXZsk:-KoyQAjmCeXue_RPXZsl:b-43icij) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_configuration_set.js#L78)):*
> ```JavaScript
>    *     configuration to be reset.
>    */
>   resetWorkspaceConfiguration(workspaceConfigurationName) {
> ```

Would this be better placed in the `WorkspaceConfigurations` class itself? Perhaps we can get the object by some get function in this class, and from there have a reset function at the `WorkspaceConfiguration` level so that each class's responsibility/scope is clearly defined.

---

*[src/model/workspace_contents_set.js, line 94 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyTxzm38jtrnvCAVzC:-KoyTxzm38jtrnvCAVzD:b-nidu42) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_contents_set.js#L94)):*
> ```JavaScript
>    *     to clear.
>    */
>   clearWorkspaceContents(workspaceContentsName) {
> ```

Maybe have this in the `WorkspaceContents` class itself?

---

*[src/view/navigation_tree.js, line 26 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQw53cBMlfwd7KI0A:-KoyQw53cBMlfwd7KI0B:bzgij0w) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/view/navigation_tree.js#L26)):*
> ```JavaScript
>  * @author sagev@google.com (Sage Vouse)
>  */
> class NavigationTree {
> ```

Would this be better named a Controller (`NavTreeController`?) if it is acting as such?
(Also if so perhaps move to the controller directory)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
comment fix

svouse committed:
fixed camel

svouse committed:
fixed a method call

svouse committed:
updated getTreeJson of individual sets

svouse committed:
made JSDoc more specific; Resource instead of Object

svouse committed:
updated TODO #100

svouse committed:
using let

svouse committed:
added TODO #103

svouse committed:
fixed merge conflicts

AnmAtAnm commented on src/controller/project_controller.js:
```diff
@@ -36,16 +36,23 @@ goog.require('Project');
     this.project = project;
+    /**
+     * The tree which represents the project.
+     * @type {!NavigationTree}
```
This says NavigationTree, but above says JTree

svouse replied:
Fixed

svouse commented:
Review status: 0 of 12 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/project_controller.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyL62c7s0-fVQL1zi1:-KoyY1doWCWkOzmewhzP:b-wcx3zp) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L91)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Good idea - another suggestion is possibly returning the object you just added (and if it wasn't successfully added, return `null`)? It may be more useful to have the object as well as an indication of whether an object was successfully added (in case there is another operation we would like to do right after adding). It has the advantages of returning a boolean in addition to more information.
</blockquote></details>

I dig it; updated the issue/comment

---

*[src/model/project.js, line 330 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyMPbX0Fl6_X2-Zcrm:-KoyYLToFVuCLb1kZKRu:b60wpuw) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/project.js#L330)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

`getLibrarySetJson()`? (here and below)

(for naming consistency, i suppose)

Also, perhaps it would be better design to put these json generating functions in their corresponding classes (i.e. instead of here, there might be a `getJson()` in every set class?).
</blockquote></details>

I dig it; the sets do indeed have getTreeJson in them, but since the set represents the organization meant to be imparted by the headings here, anyway, it makes more sense to just put them in there. Updated. (Now the methods in sets have the heading needed for project display and explicitly call the super method for children as opposed to doing the same thing in project)

---

*[src/model/resource.js, line 90 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyPBfMuQSlRScraWCH:-KoyjTU1r71ugGDw4Du-:bfuc0t5) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/resource.js#L90)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I believe JavaScript doesn't let you initialize variables as `const` if you don't give it a value upon naming it. In cases like this I'm pretty sure we have to use `let` or `var`. (Or at least my linter complains about it)
</blockquote></details>

Ah my linter was silent; I may ask you about improving it later. Updated

---

*[src/model/resource_set.js, line 57 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyPXBlOe4X1T3aQEKI:-KoybugK5j35BiODHgMB:bquhh00) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/resource_set.js#L57)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Nit - `{!Resource}`? Not sure how JSdoc works but they're all extensions of it, so maybe that works?
</blockquote></details>

SGTM. Fixed

---

*[src/model/toolbox_set.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyTttre8roj1uxTYeW:-Koyjq-N6Y7MLf0v9vXg:b-xblg0a) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/toolbox_set.js#L91)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Would it be better to have this method in the `Toolbox` class itself? 0:
</blockquote></details>

True True the class itself will need clearing capability (as well as the other Resource classes that are called); added a TODO

---

*[src/model/workspace_configuration_set.js, line 78 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQAjmCeXue_RPXZsk:-KoymF_z6trNdG-Egpv_:b6prbgd) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_configuration_set.js#L78)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Would this be better placed in the `WorkspaceConfigurations` class itself? Perhaps we can get the object by some get function in this class, and from there have a reset function at the `WorkspaceConfiguration` level so that each class's responsibility/scope is clearly defined.
</blockquote></details>

 I'm confused as to how the split you're suggesting should work; I was thinking that the set would find the object and call its reset/clear (which I agree should be in the class the set is composed of). Could you clarify?

---

*[src/model/workspace_contents_set.js, line 94 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyTxzm38jtrnvCAVzC:-KoymkLaWN7-bgQtejbf:b-v6g32m) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_contents_set.js#L94)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Maybe have this in the `WorkspaceContents` class itself?
</blockquote></details>

Same queries/suppositions as the above two

---

*[src/view/navigation_tree.js, line 26 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQw53cBMlfwd7KI0A:-Koympr_fS7nhZG17X66:b-ydrhqd) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/view/navigation_tree.js#L26)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Would this be better named a Controller (`NavTreeController`?) if it is acting as such?
(Also if so perhaps move to the controller directory)
</blockquote></details>

Good question; I think I was inclined to keep it view because it was largely just reacting to the app's data. Would you suggest in the case that it's made a controller that there's a corresponding view? I'm inclined to keep it in view, but chill either way if it works better in controller

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
updated a reference

AnmAtAnm commented on src/controller/project_controller.js:
```diff
@@ -78,87 +85,91 @@ class ProjectController {
-    // TODO: Implement
-    throw 'Unimplemented: addWorkspaceContents()';
+  //TODO #100: have component addition return object or null and have tree addition depend on it
+  addWorkspaceConfiguration(workspaceConfigName) {
```
Consider breaking out the responsibilities between creating and adding. You can add a project level convenience method `create` that unifies the two:

    projController.createWorkspaceConfiguration(name) {
      // Maybe check name conflict first.
      var wsConfig = new WorkspaceConfiguration(name);
      projController.addWorkspaceConfiguration(wsConfig);
      return wsConfig;
    }

Same for block library, etc.

rachel-fenichel commented:
Reviewed 12 of 12 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/model/resource.js, line 95 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp17NQtK2YBGpZEYb89:-Kp17NQtK2YBGpZEYb8A:b-msnu1n) ([raw file](https://github.com/google/blockly-devtools/blob/f6e67f6b6443f6fe0e03de052491c467e3bb27bc/src/model/resource.js#L95)):*
> ```JavaScript
>   makeTreeJsonFromArray(nameArray) {
>     let treeJson;
>     const iterationIndex = 1;
> ```

iterationIndex and finalIndex don't seem like const either, since you're changing them on every iteration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
comment response, TODOs

svouse commented:
Done

---

Review status: 4 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/model/resource.js, line 95 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp17NQtK2YBGpZEYb89:-Kp1BQZloJRdZNomvgMI:b-wwy2bo) ([raw file](https://github.com/google/blockly-devtools/blob/f6e67f6b6443f6fe0e03de052491c467e3bb27bc/src/model/resource.js#L95)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

iterationIndex and finalIndex don't seem like const either, since you're changing them on every iteration.
</blockquote></details>

True; fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101#-:-Kp1BNFdmm-ZhK_ZdMdV:bb6un1j)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed JSDoc in app_controller_2.js

svouse committed:
removed newline

AnmAtAnm commented:
Review status: 3 of 12 files reviewed at latest revision, 23 unresolved discussions.

---

*[src/controller/project_controller.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyL62c7s0-fVQL1zi1:-Kp1CrbFSYeYY36Lk_RL:b-rs1hq5) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L91)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

I dig it; updated the issue/comment
</blockquote></details>

Add to TODO: throw error upon naming conflict

---

*[src/controller/project_controller.js, line 92 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyKi99R0ZpWVthYuzo:-Kp1D1cgPPGT0GzbLgD8:bgmiltk) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L92)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Would it be a better idea to pass in the toolbox itself to add when adding? (As well as in the other add functions below.)

Update: Either works :D
</blockquote></details>

Per new design, this is create.  On can pass in a value via add(..) method

---

*[src/controller/project_controller.js, line 92 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1DCgSFzb6du1gDa5w:-Kp1DCgTve49pkt0Zqlv:baiso9s) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/controller/project_controller.js#L92)):*
> ```JavaScript
>    * @return {!Toolbox} The new toolbox added to the project.
>    */
>   createNewToolbox(toolboxName) {
> ```

"New" is implied by "create"

---

*[src/model/project.js, line 132 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1EggSodlhjdMoOCgF:-Kp1EggSodlhjdMoOCgG:b-fnmwux) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L132)):*
> ```JavaScript
>    //TODO #104: change param to actual object
>   addBlockLibrary(blockLibraryName) {
>     this.librarySet.addLibrary(blockLibraryName);
> ```

For future consideration:
why isn't this `addResource(..)`?  Might be able to rename this down to `.add(..)`.

---

*[src/model/project.js, line 245 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1Di7lvBv2UVsLSdlo:-Kp1Di7lvBv2UVsLSdlp:b-5w8lyb) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L245)):*
> ```JavaScript
>    * @param {string} toolboxName The name of the toolbox to clear.
>    */
>   clearToolbox(toolboxName) {
> ```

Necessary shortcut for `project.getToolbox(name).clear()`?
Adding this means more methods to maintain for very limited convenience.

Same below.

---

*[src/model/project.js, line 262 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1EBSzqZJURWSLg6bB:-Kp1EBSzqZJURWSLg6bC:bea0da7) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L262)):*
> ```JavaScript
>    *     to be reset.
>    */
>   resetWorkspaceConfiguration(workspaceConfigName) {
> ```

Similarly, `project.getWorkspaceConfiguration(name).reset()`

---

*[src/model/resource_set.js, line 65 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1F9nYYVavG3lp4tnL:-Kp1F9nYYVavG3lp4tnM:b-p6b4q9) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L65)):*
> ```JavaScript
>    * @param {!Resource} resource The resource to be added.
>    */
>   addResource(resource) {
> ```

Future PR: Consider simplifying to just `add(..)`.  We already know this is a _Resource_ Set

---

*[src/model/resource_set.js, line 73 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GDax61-r17D_VPFp:-Kp1GDax61-r17D_VPFq:b-2zw71p) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L73)):*
> ```JavaScript
>    * @param {string} resourceName The name of the resource to be removed.
>    */
>   removeResource(resourceName) {
> ```

Similarly, `remove(resourceName)`.

---

*[src/model/resource_set.js, line 82 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1G8fIm7u0XzJlQS1J:-Kp1G8fIm7u0XzJlQS1K:b34y12z) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L82)):*
> ```JavaScript
>    * @return {!Resource} The resource, or null if it's not contained in the set.
>    */
>   getResource(resourceName) {
> ```

Similarly, `get(resourceName)`

---

*[src/model/resource_set.js, line 90 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1FQ4znj-PlZITah3v:-Kp1FQ4znj-PlZITah3w:b-8c2yq4) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L90)):*
> ```JavaScript
>    * @return {Array.<string>} The names of all resources the set contains.
>    */
>   getResourceNames() {
> ```

Similarly, `getNames()`

---

*[src/model/resource_set.js, line 118 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1FxVQJ0WS06MnuW1F:-Kp1FxVQJ0WS06MnuW1G:b-l2tl1r) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L118)):*
> ```JavaScript
>    * @return {!Object} The metadata for the set.
>    */
>   getData() {
> ```

This is too vague to be in the code right now. Use case isn't clear. I think it will be deleted.

---

*[src/model/workspace_configuration_set.js, line 78 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQAjmCeXue_RPXZsk:-Kp1FbRiEIY7nDhVlVzD:bvh1x70) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_configuration_set.js#L78)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

 I'm confused as to how the split you're suggesting should work; I was thinking that the set would find the object and call its reset/clear (which I agree should be in the class the set is composed of). Could you clarify?
</blockquote></details>

Agree with Celine. No need for `configSet.reset(name)` if we can do `configSet.get(name).reset()`

---

*[src/view/navigation_tree.js, line 108 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GaH3iXatC6kvoEhb:-Kp1GaH3iXatC6kvoEhc:b-b1fysf) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L108)):*
> ```JavaScript
>    * @param {string} componentName The name of the component to add it under.
>    */
>   addBlockNode(blockType, componentName) {
> ```

Is `componentName` really `libraryName`?

---

*[src/view/navigation_tree.js, line 109 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GlSlQOTPvifixfNp:-Kp1GlSlQOTPvifixfNq:b-hqqqn1) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L109)):*
> ```JavaScript
>    */
>   addBlockNode(blockType, componentName) {
>     $('#navigationTree').jstree().create_node(componentName,
> ```

This should defer to addComponentNode, so there is a single point where this a single point to add checks or overrides, etc.

---

*[src/view/navigation_tree.js, line 110 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1HIxE9Y-d4GQLmyXD:-Kp1HIxE9Y-d4GQLmyXE:b-btnwuy) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L110)):*
> ```JavaScript
>   addBlockNode(blockType, componentName) {
>     $('#navigationTree').jstree().create_node(componentName,
>       {'id': 'block_' + blockType, 'text': blockType }, 'last', null);
> ```

Add comment about the assumption the blockType is unique across all libraries.

---

*[src/view/navigation_tree.js, line 110 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1IEcj1flhzUyaR8VF:-Kp1IEcj1flhzUyaR8VG:b6cjj32) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L110)):*
> ```JavaScript
>   addBlockNode(blockType, componentName) {
>     $('#navigationTree').jstree().create_node(componentName,
>       {'id': 'block_' + blockType, 'text': blockType }, 'last', null);
> ```

Make this prefix a constant. Similarly for others.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
comment response

svouse committed:
param fixes

svouse commented:
Review status: 3 of 12 files reviewed at latest revision, 22 unresolved discussions.

---

*[src/controller/project_controller.js, line 91 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyL62c7s0-fVQL1zi1:-Kp1IuklNwCNyUDbL-TM:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/controller/project_controller.js#L91)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add to TODO: throw error upon naming conflict
</blockquote></details>

Done.

---

*[src/controller/project_controller.js, line 92 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1DCgSFzb6du1gDa5w:-Kp1ItIwPReMsntz1tJH:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/controller/project_controller.js#L92)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

"New" is implied by "create"
</blockquote></details>

Done.

---

*[src/model/project.js, line 132 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1EggSodlhjdMoOCgF:-Kp1JDoOHOASFRJXt4UM:b-kjkm29) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L132)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

For future consideration:
why isn't this `addResource(..)`?  Might be able to rename this down to `.add(..)`.
</blockquote></details>

Agreed; added TODO to that end

---

*[src/model/project.js, line 245 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1Di7lvBv2UVsLSdlo:-Kp1KVjh4oJjB-y6zAFy:b-1ysrzd) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L245)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Necessary shortcut for `project.getToolbox(name).clear()`?
Adding this means more methods to maintain for very limited convenience.

Same below.
</blockquote></details>

I'll probably do a quick PR that handles issues #103, #104 and #106

---

*[src/model/project.js, line 262 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1EBSzqZJURWSLg6bB:-Kp1ONjdqvC0Q24iGYF7:b-bdp9h3) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/project.js#L262)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Similarly, `project.getWorkspaceConfiguration(name).reset()`
</blockquote></details>

Will do as part of the #103, #104 and #106 PR

---

*[src/model/resource_set.js, line 65 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1F9nYYVavG3lp4tnL:-Kp1OWOrfQAeQlV-LlyU:b-bdp9h3) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L65)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Future PR: Consider simplifying to just `add(..)`.  We already know this is a _Resource_ Set
</blockquote></details>

Will do as part of the #103, #104 and #106 PR

---

*[src/model/resource_set.js, line 73 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GDax61-r17D_VPFp:-Kp1OXLYrUlYcMofwI45:b-bdp9h3) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L73)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Similarly, `remove(resourceName)`.
</blockquote></details>

Will do as part of the #103, #104 and #106 PR

---

*[src/model/resource_set.js, line 82 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1G8fIm7u0XzJlQS1J:-Kp1Kx8ECg7CyMgxO80B:b-bdp9h3) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L82)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Similarly, `get(resourceName)`
</blockquote></details>

Will do as part of the #103, #104 and #106 PR

---

*[src/model/resource_set.js, line 90 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1FQ4znj-PlZITah3v:-Kp1OZp_UP3QrcgOMutZ:b-bdp9h3) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L90)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Similarly, `getNames()`
</blockquote></details>

Will do as part of the #103, #104 and #106 PR

---

*[src/model/resource_set.js, line 118 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1FxVQJ0WS06MnuW1F:-Kp1L23OQZKpLhhsNxod:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/model/resource_set.js#L118)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This is too vague to be in the code right now. Use case isn't clear. I think it will be deleted.
</blockquote></details>

Done.

---

*[src/model/workspace_configuration_set.js, line 78 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQAjmCeXue_RPXZsk:-Kp1LAcYZfdSO5cqF5FR:b-avfuxq) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/model/workspace_configuration_set.js#L78)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Agree with Celine. No need for `configSet.reset(name)` if we can do `configSet.get(name).reset()`
</blockquote></details>

Removed (and the methods will exist as part of addressing #103)

---

*[src/view/navigation_tree.js, line 108 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GaH3iXatC6kvoEhb:-Kp1LdAYzcrpm2Ta3ilU:b7m77s7) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L108)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Is `componentName` really `libraryName`?
</blockquote></details>

Yup; updated

---

*[src/view/navigation_tree.js, line 109 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1GlSlQOTPvifixfNp:-Kp1LcucLU0XPquHfZev:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L109)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This should defer to addComponentNode, so there is a single point where this a single point to add checks or overrides, etc.
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 110 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1HIxE9Y-d4GQLmyXD:-Kp1xl5RvUORyB02uOVm:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L110)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add comment about the assumption the blockType is unique across all libraries.
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 110 at r6](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp1IEcj1flhzUyaR8VF:-Kp1xmdCrR7SAny1uVBB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/62095f57a5968bc998f929dda19afd2455add604/src/view/navigation_tree.js#L110)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Make this prefix a constant. Similarly for others.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed method names called in project_controller.js

chynu commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 12 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/view/navigation_tree.js, line 26 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/101#-KoyQw53cBMlfwd7KI0A:-Kp2-84A4mJ99DH7oVj2:bvnhevm) ([raw file](https://github.com/google/blockly-devtools/blob/eb23bed72b493944b82db688942fc440f3da0768/src/view/navigation_tree.js#L26)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Good question; I think I was inclined to keep it view because it was largely just reacting to the app's data. Would you suggest in the case that it's made a controller that there's a corresponding view? I'm inclined to keep it in view, but chill either way if it works better in controller
</blockquote></details>

I guess if you move it to controller it would probably be best to split the functions into separate view/controller files. Though it would be adding more files... hm.

Since it's not too Important(TM) we can just leave it for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101#-:-Kp2-b9V73Vvoy148KXn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 2 of 12 files at r4, 3 of 8 files at r5, 1 of 2 files at r6, 3 of 5 files at r7.
Review status: 9 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/model/block_library_set.js, line 94 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp2-HYehbilDxmQJWST:-Kp2-HYehbilDxmQJWSU:b-s00sst) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/model/block_library_set.js#L94)):*
> ```JavaScript
>    */
>   //TODO #104: change param to actual object
>    addBlockToLibrary(libraryName, block) {
> ```

indent -1

---

*[src/view/navigation_tree.js, line 33 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp20ob1Te3XUE57tczU:-Kp20ob1Te3XUE57tczV:bij8dxj) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L33)):*
> ```JavaScript
>  */
> class NavigationTree {
>   /**
> ```

This comment does not refer to a single field, so use `// ..` or `/* .. */`.  `/** .. */` is only for the immediately following declaration.

---

*[src/view/navigation_tree.js, line 39 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp21GGU71qqTd04MN8b:-Kp21GGU71qqTd04MN8c:b-hvle0t) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L39)):*
> ```JavaScript
>    *     across that project.
>    */
>   toolboxPrefix = "Toolbox";
> ```

`const TOOLBOX_PREFIX = ...`
etc.

Missing BLOCK_PREFIX?

Looks like the are "node type" primarily, and "prefix" secondarily.

---

*[src/view/navigation_tree.js, line 124 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp225_F6ZE9CYlVVEvU:-Kp225_F6ZE9CYlVVEvV:b-o4kf1b) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L124)):*
> ```JavaScript
>      *     blocktypes are unique across all libraries in the project.
>      */
>     addComponentNode(libraryName, blockType);
> ```

s/libraryName/BLOCK_PREFIX/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
PR responses

svouse commented:
Review status: 9 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/model/block_library_set.js, line 94 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp2-HYehbilDxmQJWST:-Kp22Y5AUxGg3AT_sznP:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/model/block_library_set.js#L94)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

indent -1
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 33 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp20ob1Te3XUE57tczU:-Kp22hGV0A_1HzgUfqKB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L33)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This comment does not refer to a single field, so use `// ..` or `/* .. */`.  `/** .. */` is only for the immediately following declaration.
</blockquote></details>

Done.

---

*[src/view/navigation_tree.js, line 124 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp225_F6ZE9CYlVVEvU:-Kp25gqk7hPyvffEEz_8:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L124)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

s/libraryName/BLOCK_PREFIX/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/view/navigation_tree.js, line 39 at r7](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp21GGU71qqTd04MN8b:-Kp26sXrFrdOc9mS6I17:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/1393cee4dfda248104c3edd18651da6e44d53e50/src/view/navigation_tree.js#L39)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

`const TOOLBOX_PREFIX = ...`
etc.

Missing BLOCK_PREFIX?

Looks like the are "node type" primarily, and "prefix" secondarily.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

svouse committed:
fixed space

AnmAtAnm commented:
Review status: 8 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/view/navigation_tree.js, line 174 at r8](https://reviewable.io:443/reviews/google/blockly-devtools/101#-Kp27e-DZ-MKZRX77yt4:-Kp27e-DZ-MKZRX77yt5:b-z7yayg) ([raw file](https://github.com/google/blockly-devtools/blob/f50ef89818e8bd8fbfeee88e70b1388e6feab52e/src/view/navigation_tree.js#L174)):*
> ```JavaScript
>    * @param {string} parent The parent of the new node.
>    */
>   addComponentNode(prefix, componentName, parent) {
> ```

Nit `parentName`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/101#-:-Kp27oADGIq9GsMhYcsg:bnfp4nl)*
<!-- Sent from Reviewable.io -->

svouse committed:
parent => parentName
